okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -1537,21 +1540,23 @@ static int decodeHexDigit(char c) {
    *
    * @param alreadyEncoded true to leave '%' as-is; false to convert it to '%25'.
    * @param query true if to encode ' ' as '+', and '+' as "%2B".
+   * @param fragment true to leave non-ASCII codepoints as-is; false to encode them.
    */
   static String canonicalize(String input, int pos, int limit, String encodeSet,
-      boolean alreadyEncoded, boolean query) {
+      boolean alreadyEncoded, boolean query, boolean fragment) {
```

Comment:
This describes the consumption location, not the actual behavior. This method knows nothing about fragments. Maybe call it `encodeToAscii`? or `retainNonAscii`? I don't know... naming is hard.